frontend/apps/reader/modules/readeruserhyph.lua

```diff
@@ -0,0 +1,372 @@
+local DataStorage = require("datastorage")
+local Event = require("ui/event")
+local FFIUtil = require("ffi/util")
+local InfoMessage = require("ui/widget/infomessage")
+local InputDialog = require("ui/widget/inputdialog")
+local MultiConfirmBox = require("ui/widget/multiconfirmbox")
+local UIManager = require("ui/uimanager")
+local T = require("ffi/util").template
+local Utf8 = require("utf8")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local TimeVal = require("ui/timeval")
+local lfs = require("libs/libkoreader-lfs")
+local logger = require("logger")
+local _ = require("gettext")
+
+local ReaderUserHyph = WidgetContainer:new{
+    KEEP_ORIGINAL = 1, -- for dictionary file operations
+    USE_NEW = 2,       --     -"-
+}
+
+-----------------------------------------------
+--
+-----------------------------------------------
+
+-- returns path to the user dictionary
+function ReaderUserHyph:getDictionaryPath()
+    return FFIUtil.joinPath(DataStorage:getSettingsDir(),
+        self.ui.document:getTextMainLangDefaultHyphDictionary():gsub(".pattern$", "") .. "-user.hyph")
+end
+
+-- Load the user dictionary suitable for the actual language
+-- Unload is done automatically when a new dictionary is loaded.
+function ReaderUserHyph:loadDictionary()
+    if G_reader_settings:isTrue("hyph_user_dict") then
+        self.ui.document:setUserHyphenationDict(self:getDictionaryPath())
+    else
+        self.ui.document:setUserHyphenationDict()
+    end
+end
+
+-- Reload on change of the hyphenation language
+function ReaderUserHyph:onChangedLanguage()
+    self:onChangedUserDictionary()
+end
+
+-- Reload on "ChangedUserDictionary" event
+function ReaderUserHyph:onChangedUserDictionary()
+    local start_tv = TimeVal:now()
+    self.ui.document:setUserHyphenationDict(self:_checked() and self:getDictionaryPath() or "")
```

Comment:
Redundant with `loadDictionary()` above ? (I would expect to see `self.ui.document:setUserHyphenationDict()` only once, in some wrapper function in here if needed.)

Reply:
I would say, now, both do the same. But the sources of the event are totally different. The first one is called, if the hyhenation dictionary changes, the second one, if the dictionary changes its contents.
I would leave both in, as this doesn't cost much.

Reply:
I would have `ReaderUserHyph:onChangedUserDictionary()` just call `self:loadDictionary()` - if the conditions for what they do are similar (which is a bit hard for me to be sure of :/ , so I expect a bit of normalisation to make it easier to grasp :)